### ISSUE TITLE
Doesn't work when accessing from the same host as UDM Pro

### ISSUE THREAD

**Issue by edwardmp** [briis/unifiprotect#312]
Hi,

I can't seem to connect when adding the integration in HA.

In did some debugging.. and actually I encounter the same issue here as I did with another Unifi project (see https://github.com/Kane610/aiounifi/pull/70 for context)

In short, the auth endpoint is used is wrong. It uses `/api/auth` but should use `/api/auth/login`.
And this in turn is caused because `is_unifi_os ` is false, while it should be true.

Issue is that the code for determining whether the device is running UnifiOS doesn't seem to be working for all cases (maybe I use newer firmware, 1.10.4.3702?):

```python
    if response.headers.get("x-csrf-token"):
            self.is_unifi_os = True
            self.headers = {"x-csrf-token": response.headers.get("x-csrf-token")}
        else:
            self.is_unifi_os = False
```

In my case `x-csrf-token` is just not there in the response headers, causing it not be recongized as a Unifi OS device...

So we need another mechanism to identify whether the device is running Unifi OS I guess

**Comment by edwardmp** [comment]
I see #33 describes the same issue

**Comment by AngellusMortis** [comment]
Then yeah, I am not sure why it is not working. It should be. What versions of Unifi OS and Unifi Protect are you running?

Regardless, in 0.10 we will actually be removing the `is_unifi_os` check since Unifi Protect requires a Unifi OS Console now, which may solve your issue. We do need to figure out why you are not getting the CSRF token though since that will likely cause other problems if you should be getting it.

You can try editing `unifiprotect/manifest.json` to replace `pyunifiprotect==0.33.1` to `git+https://github.com/AngellusMortis/pyunifiprotect.git@feature/remove-non-unifios#pyunifiprotect==0.33.1`. Then restart HA.

That will give you this version of pyunifiprotect: https://github.com/briis/pyunifiprotect/pull/54

**Comment by edwardmp** [comment]
Well so the reason it doesn't serve the CSRF header is because the app uses UnifiProtect runs on the Dream Machine Pro itself (through Docker). Apparently, therefore the Dream machine software doesn't serve the header.

I'll try your new branch, I guess that should indeed fix my issue going forward.

**Comment by AngellusMortis** [comment]
The branch has been merged now. So you will need the latest beta of Unifi Protect (1.20-beta5) and the latest HA Integration (0.10 beta3) and it should have the code already.

**Comment by edwardmp** [comment]
Sweet! Will report back to you asap

**Comment by AngellusMortis** [comment]
Also, you are running Unifi Protect yourself on the UDM Pro? Via docker? It is not being ran by the UDM Pro podman `unifi-so` container?

**Comment by edwardmp** [comment]
Yeah! Have an UDM Pro and just running protect on the host OS itself

**Comment by AngellusMortis** [comment]
Do you have a guide to how you are running Protect by itself? UnifiOS should be managing it for you. On my UDM Pro it runs inside of the podman container `unifi-so`.

Also what port are you using connect to Protect? Normally on a UnifiOS console, UnifiOS proxies Protect through port 443 and I am guess that is when it adds the CSRF token.

**Comment by edwardmp** [comment]
I just checked, I do have a running container with name `unifi-os`. Otherwise, it's just a completely default setup where I installed/configured Protect through the console.

But so the issue is that I'm running HA also through docker on the dream machine itself, so there must be some logic somewhere to not add the CSRF token for local requests.

**Comment by AngellusMortis** [comment]
Ah okay. So yeah, then you are running into the same exactly issue as https://github.com/briis/pyunifiprotect/issues/33 then. Which was my initial suspicion. 

And I do not know a way to manage HACS integrations through command line. You could manually clone the repo, but you need to make sure you upgrade Unifi Protect to the latest 1.20 beta first as the 0.10 HA integration requires it going forward.

If you installed the integration through HACS, use HACS to upgrade it. Otherwise, it could cause some issues. If you installed it manually, you should be able to do something like this to install it:

```bash
cd /path/to/HA/data/volume
git clone https://github.com/briis/unifiprotect.git
cd custom_components
ln -s ../unifiprotect/custom_components/unifiprotect unifiprotect
```

Then restart HA.

**Comment by edwardmp** [comment]
Duplicate of #33

**Comment by AngellusMortis** [comment]
Steps to install the latest Protect beta as well in case you have never done it:

1. Log in as the UDM Pro owner (via your UI Cloud account)
2. Go to the System Settings on the UnifiOS Portal.
3. Go to Advanced and make sure Remote Access is enabled (you can disable it after upgrading to the beta)
4. Go to Unifi Network (New UI) -> Settings -> System -> Maintenance  -> Enable Private Release Channels and set Unifi Device Update Release Channel to Early Access
5. Go back to the UnifiOS Portal System settings.
6. Updates -> Wait until Early Access is a choice in the release channel drop down and change it.
7. Now you should get the update for Protect. 
8. After Protect is fully updated and on the latest beta, you can go back and disable Remote Access

**Comment by edwardmp** [comment]
Yeah thanks already figured out how to do that.

Unfortunately, after installing this integration through HACS again from master branch, I still get:

> Error retrieving data from Unifi Protect.

Logs:
>. 2021-10-30 13:59:30 DEBUG (MainThread) [custom_components.unifiprotect.config_flow] Call https://192.168.1.1:443/api/auth received 404 Not Found

When configuraring the integration

**Comment by AngellusMortis** [comment]
It looks like you might not running the latest 0.10 beta. Make sure you have Show Beta versions check.

![image](https://user-images.githubusercontent.com/490848/139536127-f2c6b403-8ff6-450c-9c61-fc880fa3a34d.png)

**Comment by AngellusMortis** [comment]
A way to check what version of pyunifiprotect you have install is to exec into your HA docker container and then run

```bash
pip list | grep pyunifiprotect
```

You should have version `0.35.0`. The code that was added that should help in your case was added in 0.35.0 (which comes from beta3) you so _have_ to have that beta/version.

**Comment by edwardmp** [comment]
Ah indeed, I had to uninstall protect from HACS one more time, last time even with the beta version switch toggled it didn't show up.

This time an unknown error:

```
File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 63, in wrapper
result = await method(view, request, *args, **kwargs)
File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 109, in post
result = await self._flow_mgr.async_configure(flow_id, data)
File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 202, in async_configure
result = await self._async_handle_step(flow, cur_step["step_id"], user_input)
File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 260, in _async_handle_step
result: FlowResult = await getattr(flow, method)(user_input)
File "/config/custom_components/unifiprotect/config_flow.py", line 66, in async_step_user
server_info = await unifiprotect.server_information()
File "/config/deps/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 376, in server_information
return await self._get_server_info()
File "/config/deps/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 386, in _get_server_info
data = await self.api_request("bootstrap")
File "/config/deps/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 166, in api_request
response = await self.request(method, url, require_auth=False, auto_close=False, **kwargs)
File "/config/deps/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 125, in request
response = await req_context.__aenter__()
File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 1117, in __aenter__
self._resp = await self._coro
File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 542, in _request
resp = await req.send(conn)
File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 668, in send
await writer.write_headers(status_line, self.headers)
File "/usr/local/lib/python3.9/site-packages/aiohttp/http_writer.py", line 118, in write_headers
buf = _serialize_headers(status_line, headers)
File "aiohttp/_http_writer.pyx", line 137, in aiohttp._http_writer._serialize_headers
File "aiohttp/_http_writer.pyx", line 109, in aiohttp._http_writer.to_str
TypeError: Cannot serialize non-str key None
```

it's probably the same issue that I described in the linked issue of that another Unifi link: since the CSRF header is not there, it has a value of None which can't be serialized to a string.

I'll make a local change and open a pull request for that when it works

**Comment by AngellusMortis** [comment]
Yep, that is exactly the case. That is what I was suspecting. I am actually going to make a PR now.

**Comment by AngellusMortis** [comment]
Made the PR: https://github.com/briis/pyunifiprotect/pull/67.

You should be able to use it in your HA instance by editing `custom_components/unifiprotect/manifest.json` and replacing `pyunifiprotect==0.35.0` with `git+https://github.com/AngellusMortis/pyunifiprotect.git@feature/no-csrf-token#pyunifiprotect==0.35.3`

**Comment by edwardmp** [comment]
Nice, exact same [change[(https://github.com/Kane610/aiounifi/pull/70/commits/ca415332deea8c3efa4cd3b830b7220fb7f55fc1) as I did for this other lib, history repeats

**Comment by AngellusMortis** [comment]
Just let me know if it works, then I will mark the PR ready and get @briis to merge it.

**Comment by edwardmp** [comment]
Yes, after changing the manifest the change was not applied (do I need to do something else there)?

But I just manually applied your patch locally, and now it works fine. Nice one! 😄

**Comment by AngellusMortis** [comment]
This should be fixed in the 0.10 RC. Let us know if you have any other issues.
